tests/microservices/Microservices.FileCopier.Tests/Messaging/FileCopyQueueConsumerTest.cs

```diff
@@ -0,0 +1,184 @@
+        }
+
+        [TearDown]
+        public void TearDown() { }
+
+        private static BasicDeliverEventArgs GetMockDeliverArgs(ExtractFileMessage message)
+        {
+            var mockDeliverArgs = Mock.Of<BasicDeliverEventArgs>(MockBehavior.Strict);
+            mockDeliverArgs.DeliveryTag = 1;
+            mockDeliverArgs.Body = Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(message));
+            mockDeliverArgs.BasicProperties = new BasicProperties { Headers = new Dictionary<string, object>() };
+            var header = new MessageHeader();
+            header.Populate(mockDeliverArgs.BasicProperties.Headers);
+            // Have to convert these to bytes since RabbitMQ normally does that when sending
+            mockDeliverArgs.BasicProperties.Headers["MessageGuid"] = Encoding.UTF8.GetBytes(header.MessageGuid.ToString());
+            mockDeliverArgs.BasicProperties.Headers["ProducerExecutableName"] = Encoding.UTF8.GetBytes(header.ProducerExecutableName);
+            mockDeliverArgs.BasicProperties.Headers["Parents"] = Encoding.UTF8.GetBytes(string.Join("->", header.Parents));
+            return mockDeliverArgs;
+        }
+
+        #endregion
+
+        #region Tests
+
+        [Test]
+        public void Test_FileCopyQueueConsumer_ValidMessage_IsAcked()
+        {
+            BasicDeliverEventArgs mockDeliverArgs = GetMockDeliverArgs(_message);
+
+            var consumer = new FileCopyQueueConsumer(_mockFileCopier.Object);
+            consumer.SetModel(_mockModel.Object);
+
+            var fatalCalled = false;
+            FatalErrorEventArgs fatalErrorEventArgs = null;
+            consumer.OnFatal += (sender, args) =>
+            {
+                fatalCalled = true;
+                fatalErrorEventArgs = args;
+            };
+
+            consumer.ProcessMessage(mockDeliverArgs);
+
+            Thread.Sleep(500); // Fatal is race-y
+            Assert.False(fatalCalled, $"Fatal was called with {fatalErrorEventArgs}");
+            Assert.AreEqual(1, consumer.AckCount);
+            Assert.AreEqual(0, consumer.NackCount);
+        }
+
+        [Test]
+        public void Test_FileCopyQueueConsumer_ApplicationException_IsNacked()
+        {
+            BasicDeliverEventArgs mockDeliverArgs = GetMockDeliverArgs(_message);
+
+            _mockFileCopier.Reset();
+            _mockFileCopier.Setup(x => x.ProcessMessage(It.IsAny<ExtractFileMessage>(), It.IsAny<IMessageHeader>())).Throws<ApplicationException>();
+
+            var consumer = new FileCopyQueueConsumer(_mockFileCopier.Object);
+            consumer.SetModel(_mockModel.Object);
+
+            var fatalCalled = false;
+            FatalErrorEventArgs fatalErrorEventArgs = null;
+            consumer.OnFatal += (sender, args) =>
+            {
+                fatalCalled = true;
+                fatalErrorEventArgs = args;
+            };
+
+            consumer.ProcessMessage(mockDeliverArgs);
+
+            Thread.Sleep(500); // Fatal is race-y
```

Comment:
Not a big fan of these Sleep(X), is there a reason not to use `TestTimelineAwaiter`

Reply:
Not a fan either to be honest, but I wasn't sure that TestTimelineAwaiter would work in this case. I'll take another look.

Reply:
Let me know if theres an issue with the awaiter since we should have one way of handling 'block until x' and it should be a joy to use* so if we can do anything to simplify it or make it more elegant then let me know.

*for certain small values of joy

Reply:
I've updated this now to remove the sleeps. For the cases where we don't expect Fatal to be called we can just check AckCount/NackCount. Previously I had the Fatal check there because the logs weren't being captured so the failure reason couldn't be seen. I've fixed this by moving the Logger.Fatal call into the Consumer class. For the cases where we _do_ expect Fatal to be called, we can simply use the TestTimelineAwaiter.